flink-core/src/main/java/org/apache/flink/api/dag/Transformation.java

```diff
@@ -136,6 +138,14 @@ public static int getNewNodeId() {
 	 */
 	private ResourceSpec preferredResources = ResourceSpec.DEFAULT;
 
+	/**
+	 * This weight indicates how much this transformation relies on managed memory, so that
+	 * transformation highly relies on managed memory would be able to acquire more managed
+	 * memory in runtime (linear association). Note that it only works in cases of UNKNOWN
+	 * resources.
+	 */
+	private int managedMemoryWeight = DEFAULT_MANAGED_MEMORY_WEIGHT;
```

Comment:
how to express `I don't need managed memory`?

Reply:
You can set the weight to value 0 explicitly.

Reply:
How about making `don't need managed memory` as default?

Reply:
This is to be aligned with other resources like cpu and heap memory. Operators with UNKNOWN resources are always able to acquire all kinds of available resources currently. 
I'd prefer to not making managed memory a special one.

Reply:
So what kind of ResourceSpec I should use when some operator doesn't need managed memory? First set UNKNOWN resource and then set managed memory to 0 explicitly?

Reply:
I think a valid approach would be to set the environment wide default resources to be something other than UNKNOWN, so that after stream graph is built, users (including sql) would not need to check which resources/weight is not specified and needs to be overriden with a user wanted default value (what is the current case in ExecutorUtils#setBatchProperties). 
However, we cannot take this approach for now since resources setting is not open to external users yet.

Reply:
> So what kind of ResourceSpec I should use when some operator doesn't need managed memory? First set UNKNOWN resource and then set managed memory to 0 explicitly?

Yes. Weights only works in cases of UNKNOWN resources.

Reply:
I am +1 to default `don't need managed memory`.
Note it is manage memory, only way is requiring memory from memory manager explicitly. It is different from other resources.

And image user write an operator which using manage memory, and he believe there is only one operator to use manage memory, that is his operator. But if he use DataStream/DataSet api, whatever operators including map/source... these operators will rob his memories.

Reply:
> I am +1 to default `don't need managed memory`.
> Note it is manage memory, only way is requiring memory from memory manager explicitly. It is different from other resources.
> 
> And image user write an operator which using manage memory, and he believe there is only one operator to use manage memory, that is his operator. But if he use DataStream/DataSet api, whatever operators including map/source... these operators will rob his memories.

RocksDB StateBackends also use managed memory. Making it 0 by default would break existing streaming jobs if users do not explicitly set the managed memory.

Reply:
>> Note it is manage memory, only way is requiring memory from memory manager explicitly. It is different from other resources.

How much managed memory an operator can acquire actually depends on the declared resources rather the the needed resources. For example, one can specify the task_heap_memory/task_offheap_memory/managed_memory of an operator to be a large number even if the operator does not use that much or even that kinds of resources. 
The framework should respect that settings. And I don't see managed memory to be special here.

Reply:
>> And image user write an operator which using manage memory, and he believe there is only one operator to use manage memory, that is his operator. But if he use DataStream/DataSet api, whatever operators including map/source... these operators will rob his memories.

The weight is not a public interface and users cannot set it. So if the user uses the fraction, 0 default value will always result in 0 managed memory fraction, even if the operator requires managed memory.

Reply:
I agree with @zhuzhurk that the default value should not be `0`. Otherwise we have a problem if a user writes a stateful `DataStream` program using RocksDB as he cannot set the weight value. And also if he could, then he would need to remember to set it otherwise his operator wouldn't get any managed memory.